tests/switch.t

```diff
@@ -0,0 +1,144 @@
+local test = require 'test'
+
+function dotest(fn, a, b, c, d)
+  test.eq(fn(1), a)
+  test.eq(fn(2), b)
+  test.eq(fn(3), c)
+  test.eq(fn(4), d)
+end
+
+terra foo(a: int): int
+  switch a do
+    case 1 then
+      return 2
+    case 2 then
+      return 5
+    end
+  else
+    return 3
+  end
+end
+
+dotest(foo, 2, 5, 3, 3)
+
+terra empty(a: int): int
+  switch a do
+  end
+  return -1
+end
+
+dotest(empty, -1, -1, -1, -1)
+
+terra single(a: int): int
+  switch a do
+    case 1 then 
+      return 2
+    end
+  end
+  return -1
+end
+
+dotest(single, 2, -1, -1, -1)
+
+terra justelse(a: int): int
+  switch a do
+    else
+      return 3
+  end
+  return -1
+end
+
+dotest(justelse, 3, 3, 3, 3)
+
+terra values(a: int): int
+  switch a + (a + a) do
+    case 1 + (1 + 1) then
```

Comment:
What's going on here? Do we do constant folding, or is this somehow implicitly escaping a Lua expression?

Reply:
LLVM does it's own constant folding. The only requirement we have is that whatever LLVM IR is emitted can be folded into a integer constant from the perspective of LLVM. Hence, when this compiles, LLVM sees the constants, folds them itself, and it ends up working.